Defer Unit Conversion for WELPI Values

bska opened the pull request:
This PR switches the WELPI scaling factor protocol to requiring that the caller convert the effective PI value to output units before calling
```
Well::getWellPIScalingFactor()
```
This clearly makes the function less convenient to use, but is on the other hand the only way to guarantee that client code has a consistent view of the "preferred" phase for injectors.  PR #2018's approach of capturing the preferred phase at the time of processing `WELPI` was, while well-intentioned, based on a misunderstanding of the way these keywords interact.  Moreover, my [answer](https://github.com/OPM/opm-common/pull/2018#issuecomment-707997529) to @joakim-hove's [question](https://github.com/OPM/opm-common/pull/2018#issuecomment-707991480) in turn was based on an incomplete understanding of what goes into the mobility term of the PI calculation.

The upside of this revised implementation is that we no longer need the more elaborate `WellProductivityIndex` structure to have enough information to correctly calculate the WELPI-based scaling factor.  We're therefore able to revert the earlier commit 3eef45e and remove all traces of that structure.

Finally record a general, `Schedule`-wide `WELL_PRODUCITIVITY_INDEX` event in addition to the well-specific events we already record when processing a WELPI keyword.  We need this information to trigger a PI calculation across all MPI processes in opm-simulators.

bska commented:
jenkins build this please

joakim-hove commented:
OK - this looks fine in itself; but how will we guarantee the correct units when using this?

bska commented:
> how will we guarantee the correct units when using this?

There is sadly no way to guarantee that.  Client code must use something along the lines of the snippet below when calling `Well::getWellPIScalingFactor()`.

---

```C++
double effectivePI = 0.0;
switch (well.getPreferredPhase()) {
case Phases::OIL:
    effectivePI = unit_system.from_si(measure::liquid_productivity_index, computeOIlPI(args));
    break;

case Phases:GAS:
    effectivePI = unit_system.from_si(measure::gas_productivity_index, computeGasPI(args));
    break;

case Phases::WATER:
    effectivePI = unit_system.from_si(measure::liquid_productivity_index, computeWaterPI(args));
    break;

default:
    do_something();
}

double scalingFactor = well.getWellPIScalingFactor(effectivePI);
```
I suppose we could push the actual unit conversion into `Well::getWellPIScalingFactor()` if we add a `UnitSystem` parameter to the function, but calling code will still have to look at `Well::getPreferredPhase()` to know which PI value to pass as an argument.

bska commented:
I'm putting this back to work-in-progress status because I've rebased the contents onto the top of #2068 and we should not merge this before that other PR is installed in the master branch.  I'm nevertheless going to do a new build check.

bska commented:
jenkins build this please

joakim-hove commented:
> I suppose we could push the actual unit conversion into Well::getWellPIScalingFactor() if we add a UnitSystem parameter to the function, but calling code will still have to look at Well::getPreferredPhase() to know which PI value to pass as an argument.

My suggestion would be to only calculate in SI in the simulator; and then pass preferred phase along with the SI number to `Well:getWillPIScalingFactor()` - the `Well?`object already has a unit system member (or at least one close by).

bska commented:
> the `Well` object already has a unit system member (or at least one close by).

Interesting.  I've been looking for a way to access `UnitSystem` information from `Schedule` contexts and I've never been successful.  I've always had to somehow conjure up an `EclipseState` object in order to call [`EclipseState::getUnits()`](https://github.com/OPM/opm-common/blob/8d3e4191626b65e23656e7237e4f76c0782d7920/opm/parser/eclipse/EclipseState/EclipseState.hpp#L110-L111).  

> pass preferred phase along with the SI number to `Well::getWillPIScalingFactor()`

The `Well` already knows its own preferred phase.  We don't need to pass that as an argument.

joakim-hove commented:
> Interesting. I've been looking for a way to access UnitSystem information from Schedule contexts and I've never been successful. 

OK - I have double checked now; the `Well` class has `UnitySystem` member. If it is needed in the `Schedule` object itself w e can add it there as well!

bska commented:
> > Interesting. I've been looking for a way to access `UnitSystem` information from `Schedule` contexts and I've never been successful.
> 
> OK - I have double checked now; the `Well` class has `UnitySystem` member.

Indeed it does.  That certainly solves the current conundrum.  Thanks!  I wonder why I didn't see it before...

> If [`UnitSystem` support] is needed in the `Schedule` object itself we can add it there as well!

I don't have an immediate need for that now, but I'll keep this option in mind if the need does arise in the future.

bska commented:
jenkins build this please

bska commented:
I have rebased this onto current master since #2068 was merged.  I have also switched the `Well::getWellPIScalingFactor()` function back to expecting an input argument in strict SI unit.  I'm marking this as "ready for review" and I'm also running a new build check.

bska commented:
jenkins build this please
